cli/azd/pkg/account/subscriptions_manager.go

```diff
@@ -173,43 +187,66 @@ func (m *SubscriptionsManager) ListSubscriptions(ctx context.Context) ([]Subscri
 		return nil, fmt.Errorf("listing tenants: %w", err)
 	}
 
-	allSubscriptions := []Subscription{}
-	errors := []error{}
-	oneSuccess := false
-
-	for _, tenant := range tenants {
-		tenantId := *tenant.TenantID
-		subscriptions, err := m.service.ListSubscriptions(ctx, tenantId)
-		if err != nil {
-			errorMsg := err.Error()
-			displayName := *tenant.DisplayName
-			if strings.Contains(errorMsg, "AADSTS50076") {
-				errors = append(
-					errors,
-					fmt.Errorf(
+	span.SetAttributes(fields.AccountSubscriptionsListTenantsFound.Int(len(tenants)))
+
+	listForTenant := func(
+		jobs <-chan armsubscriptions.TenantIDDescription,
+		results chan<- tenantSubsResult,
+		service *azcli.SubscriptionsService) {
+		for tenant := range jobs {
+			azSubs, err := service.ListSubscriptions(ctx, *tenant.TenantID)
+			if err != nil {
+				errorMsg := err.Error()
+				displayName := *tenant.DisplayName
+				if strings.Contains(errorMsg, "AADSTS50076") {
+					err = fmt.Errorf(
 						"%s requires Multi-Factor Authentication (MFA). "+
 							"To authenticate, login with `azd login --tenant-id %s`",
 						displayName,
-						*tenant.DefaultDomain))
-			} else {
-				errors = append(
-					errors,
-					fmt.Errorf("failed to load subscriptions from tenant '%s' : %w", displayName, err))
+						*tenant.DefaultDomain)
+				} else {
+					err = fmt.Errorf("failed to load subscriptions from tenant '%s' : %w", displayName, err)
+				}
 			}
 
+			results <- tenantSubsResult{toSubscriptions(azSubs, *tenant.TenantID), err}
+		}
+	}
+
+	numJobs := len(tenants)
+	jobs := make(chan armsubscriptions.TenantIDDescription, numJobs)
+	results := make(chan tenantSubsResult, numJobs)
+	// Apply max number of concurrent workers at 50
+	numWorkers := int(math.Min(float64(len(tenants)), 50))
+	for i := 0; i < numWorkers; i++ {
+		go listForTenant(jobs, results, m.service)
+	}
+
+	for i := 0; i < numJobs; i++ {
+		jobs <- tenants[i]
+	}
+	close(jobs)
```

Comment:
There is an opportunity there to move this fanout logic out into its own package.  Maybe part of the `async` package so that we could more easily apply this fan-out pattern to other parts of the CLI?

Reply:
Let's do this when we have enough recurring usage to warrant it. Implementing simple fan-in, fan-out will increase abstraction and reuse, but at the same time, reduce flexibility of creating more complex concurrent programs that may be needed.